lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -2060,4 +1917,270 @@ namespace Js
         }
         return returnValue;
     }
+
+    /*static*/
+    template <bool tryCopy, bool assign>
+    void JavascriptObject::CopyDataPropertiesHelper(Var source, RecyclableObject* to, ScriptContext* scriptContext, PropertyId* excluded, uint32 excludedLength)
+    {
+        // Note: If assign is true, excluded and excludedLength are ignored. Otherwise, tryCopy is ignored.
+        // 1. Assert Type(to) is Object.
+        // 2. Assert Type(excluded) is List.
+        // 3. If source is undefined or null, let keys be a new empty List
+        // 4. Else,
+        //    a. Let from be ToObject(source).
+        //    b. Let keys be from.[[OwnpropertyKeys]]().
+        // 5. Repeat for each element nextKey of keys in List order,
+        //    a. If we are not in assign mode, then
+        //        i. Let found be false.
+        //        ii. Repeat for each element e of excluded,
+        //            1. If e is not empty and SameValue(e, nextKey) is true, then
+        //                a. Set found to true.
+        //    b. If found is false or we are in assign mode, then
+        //        i. Let desc be from.[[GetOwnProperty]](nextKey).
+        //        ii. If desc is not undefined and desc.[[Enumerable]] is true, then
+        //            1. Let propValue be Get(from, nextKey).
+        //            2. If we are in assign mode, then
+        //                a. Perform Set(to, nextKey, propValue, true).
+        //            3. Else,
+        //                a. Perform CreateDataProperty(to, nextKey, propValue).
+        // 6. Return target (target is returned in pointer).
+        Assert(JavascriptOperators::IsObject(to));
+        RecyclableObject* from = nullptr;
+        if (!JavascriptConversion::ToObject(source, scriptContext, &from))
+        {
+            if (JavascriptOperators::IsUndefinedOrNull(source))
+            {
+                return;
+            }
+            if (assign)
```

Comment:
We now have this if/else pattern for throwing type errors several places. Could it maybe go in a helper, so we could condense this back to something like `JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_NeedObject, GetCopyDataPropertiesEntryName<assign>());`? This has the side benefit of not having repeated copies of strings like "Object.assign", so someone who wants to change all of these messages at once has to only change one line.